File: src/main/kotlin/_2Classes.java

class PersonJ {
    private String name;
    private int age;

    public PersonJ(String name) {
        this.name = name;
    }

    public PersonJ() {
        this.name = "default name";
    }

    void printNameLength() {
        System.out.println(name.length());
    }

    //Breaking constructor
    public PersonJ(int age) {
        this.age = age;
    }
}

File: src/main/kotlin/_2Classes.kt

open class Person (val name: String) {

    init {
        println("Init block")
    }

    constructor(name: String, age: Int) : this(name) {
        println("Secondary constructor")
    }

    constructor() : this("default name") {
        println("Secondary constructor")
    }

    private fun sayHello() {
        println("Hello, $name")
    }

    companion object {
        fun createPerson(name: String): Person {
            val person = Person(name)
            person.sayHello() //private function can be called from the companion object
            return person
        }
    }
}


fun main() {
    val person = Person.createPerson("John")
    println(person.name)
}
